kyo-core/shared/src/main/scala/kyo/layers.scala

```diff
@@ -0,0 +1,84 @@
+package kyo
+
+import Layers.internal.*
+import kyo.Flat.unsafe
+import kyo.core.*
+import scala.annotation.targetName
+
+sealed trait Layer[-In, +Out, -S]:
+    self =>
+
+    def >>>[In2, Out2, S2](that: Layer[Out & In2, Out2, S2]): Layer[In & In2, Out2, S & S2]       = To(self, that)
+    def ++[In2, Out2, S2](that: Layer[In2, Out2, S2]): Layer[In & In2, Out & Out2, S & S2]        = And(self, that)
+    def >+>[In2, Out2, S2](that: Layer[Out & In2, Out2, S2]): Layer[In & In2, Out & Out2, S & S2] = self ++ (self >>> that)
+
+    private[kyo] def doRun(
+        memoMap: scala.collection.mutable.Map[Layer[?, ?, ?], Any] = scala.collection.mutable.Map.empty[Layer[?, ?, ?], Any]
```

Comment:
Kyo provides stronger guarantees regarding purity across transformations of a computation. If an effect in a layer initialization invokes its continuation multiple times, this map can break referential transparency. Could this be a recursive method instead using an immutable `Map`?

Reply:
Kit and I tried a recursive encoding and weren't able to make it work. Would you prefer immutable map + atomic?

Reply:
I'll give it a try later today

Reply:
Sorry for the delay, I took a look but I can't see a way to avoid mutability for memoization. What do you think about adding `IOs` suspensions for the parts of the code that interact with the mutable map? It'd address the referential transparency issue but `Layer.run` would depend on `IOs`. Also, don't we need a concurrent map here since layers might be initialized in parallel?

Reply:
As it stands today, we don't use any par related operators. That is something we could add in the future, but not sure if the complexity is warranted now.

A dependency on IOs should be alright, it will just remove the ability to use layers as a general memoization technique... IMO it's not too big of a deal as it's pretty dependency injection specific